compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -985,7 +985,180 @@ OMR::Power::TreeEvaluator::vstoreiEvaluator(TR::Node *node, TR::CodeGenerator *c
+     case TR::Int32:
+       return TR::TreeEvaluator::vreductionAddWordHelper(node,cg, type);
+     case TR::Int64:
+       return TR::TreeEvaluator::vreductionAddDoubleWordHelper(node,cg, type);
+     case TR::Float:
+       return TR::TreeEvaluator::vreductionAddWordHelper(node,cg, type);
+     case TR::Double:
+       return TR::TreeEvaluator::vreductionAddDoubleWordHelper(node,cg, type);
+     default:
+       TR_ASSERT_FATAL(false, "unrecognized vector type %s\n", firstChild->getDataType().toString()); return NULL;
+     }
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::vreductionAddSubWordHelper(TR::Node *node, TR::CodeGenerator *cg, TR::InstOpCode::Mnemonic sumOp)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+
+   TR::Register *srcReg = cg->evaluate(firstChild);
+   TR::Register *resReg = cg->allocateRegister();
+
+   TR::Register *tempRes = cg->allocateRegister(TR_VRF);
+   TR::Register *zeroReg = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(resReg);
+
+   //Evaluate sum across vector operand. Since the operands will be either 8 or 16 bytes long, and thus the results cannot overflow, it is safe to use saturate sum instructions here
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, zeroReg, 0);
+   generateTrg1Src2Instruction(cg, sumOp, node, tempRes, srcReg, zeroReg); //Sum across each word element
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vsumsws, node, tempRes, tempRes, zeroReg); //Sum word elements across entire vector operand
+
+   //Copy result into GPR
+   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P9))
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrld, node, resReg, tempRes);
+   else
+   {
+      generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::xxpermdi, node, tempRes, tempRes, zeroReg, 3); //move sum to upper doubleword element of tempRes
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrd, node, resReg, tempRes);
+   }
+
+   cg->stopUsingRegister(tempRes);
+   cg->stopUsingRegister(zeroReg);
+   cg->decReferenceCount(firstChild);
+
+   return resReg;
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::vreductionAddWordHelper(TR::Node *node, TR::CodeGenerator *cg, TR::DataType type)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+
+   TR::Register *srcReg = cg->evaluate(firstChild);
+   TR::Register *resReg;
+
+   TR::Register *shiftReg = cg->allocateRegister(TR_VRF);
+   TR::Register *temp = cg->allocateRegister(TR_VRF);
+   TR::Register *tempGPR;
+   TR::Register *tempVSX;
+
+   //choose addOp and allocate registers based on data type
+   TR::InstOpCode::Mnemonic addOp;
+
+   if (type == TR::Int32)
+   {
+      addOp = TR::InstOpCode::vadduwm;
+      tempVSX = cg->allocateRegister(TR_VRF);
+      tempGPR = cg->allocateRegister();
+      resReg = tempGPR;
+   }
+   else if (type == TR::Float)
+   {
+      addOp = TR::InstOpCode::xvaddsp;
+      tempVSX = cg->allocateRegister(TR_FPR);
+      resReg = tempVSX;
```

Comment:
How about adding the following line here?

```
tempGPR = NULL; // Not used for TR::Float
```

Otherwise you can do it at the variable declaration line above.
```
TR::Register *tempGPR = NULL;
```